core/src/core/transaction.rs

```diff
@@ -36,6 +39,168 @@ use util::static_secp_instance;
 use util::RwLock;
 use util::ToHex;
 
+/// Fee fields as in fix-fees RFC: { future_use: 20, fee_shift: 4, fee: 40 }
+#[derive(Debug, Clone, Copy, PartialEq)]
+pub struct FeeFields(u64);
+
+impl DefaultHashable for FeeFields {}
+
+impl Writeable for FeeFields {
+	fn write<W: Writer>(&self, writer: &mut W) -> Result<(), ser::Error> {
+		writer.write_u64(self.0)
+	}
+}
+
+impl Readable for FeeFields {
+	fn read<R: Reader>(reader: &mut R) -> Result<Self, ser::Error> {
+		let fee_fields = reader.read_u64()?;
+		Ok(Self(fee_fields))
+	}
+}
+
+impl Display for FeeFields {
+	fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+		write!(f, "{}", self.0)
+	}
+}
+
+impl Serialize for FeeFields {
+	fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
+	where
+		S: Serializer,
+	{
+		serializer.collect_str(&self.0)
+	}
+}
+
+impl<'de> Deserialize<'de> for FeeFields {
+	fn deserialize<D>(deserializer: D) -> Result<FeeFields, D::Error>
+	where
+		D: Deserializer<'de>,
+	{
+		struct FeeFieldsVisitor;
+		impl<'de> de::Visitor<'de> for FeeFieldsVisitor {
+			type Value = FeeFields;
+
+			fn expecting(&self, formatter: &mut fmt::Formatter) -> fmt::Result {
+				formatter.write_str("an 64-bit integer")
+			}
+
+			fn visit_str<E>(self, value: &str) -> Result<Self::Value, E>
+			where
+				E: de::Error,
+			{
+				let value = value
+					.parse()
+					.map_err(|_| E::custom(format!("invalid fee field")))?;
+				self.visit_u64(value)
+			}
+
+			fn visit_u64<E>(self, value: u64) -> Result<Self::Value, E>
+			where
+				E: de::Error,
+			{
+				FeeFields::try_from(value).map_err(|_| E::custom(format!("invalid fee field")))
+			}
+		}
+
+		deserializer.deserialize_any(FeeFieldsVisitor)
+	}
+}
+
+/// Conversion from a valid fee to a FeeFields with 0 fee_shift
+/// The valid fee range is 1..FEE_MASK
+impl TryFrom<u64> for FeeFields {
```

Comment:
Would it be beneficial to also have a `From<u32> for FeeFields` that never fails?
I'm not sure if this simplifies any use of this but it would potentially get rid of some uses of `unwrap()`.
We'd probably need to convert to `u64 -> u32` safely though, so maybe it just moves the problem around.

Reply:
You can only get rid of unwrap by limiting fees to 32 bits, but if that was a good idea, then the RFC should have used 32 bits instead of 40 bits. This would come back to bite us in the distant future if blocks fill up and people do want to pay 5 Grin in fees, but either their wallet won't let them, or worse, will fail inexplicably.

Reply:
No I mean you can always fit a 32 bit fee safely in the 40 available bits. Not to limit fees to 32 bits.
So have the ability to turn a u32 safely into a FeeFields via `into()` (for convenience) and a u64 via `try_into()`.

Reply:
So this would only be used in tests? I thought you were suggesting a new method that wallets could use.
For literal constant fees, it would be fine. But personally, I don't mind the unwrap()s.